### ISSUE TITLE
How do I properly dimensionalize my array and tune `rnn.graph.unroll` to make the LSTM work for this multidimensional sequence

### ISSUE THREAD

**Issue by alexmosc** [apache/incubator-mxnet#12002]
It is essentially a call for help rather than code-related issue.

Assume a matrix with 5 rows and 20 columns. Each column is a sample of a multivariate timeseries. Each row is one dimension of the multivariate timeseries.

I have also a vector of 20 output values.

I am trying to build an LSTM model with sequence length = 20 which would iterate over samples 1 to 20 and regress output values associated.

I get all sorts of "shape mismatch" and "You are trying to split the 0-th axis of input tensor with shape" error messages. 

The question is how I properly dimensionalize my array of input data and tune `rnn.graph.unroll` to make the LSTM work for this multidimensional sequence.

```
library(mxnet)

rm(symbol)

symbol <- rnn.graph.unroll(seq_len = 20, 
                           num_rnn_layer =  1, 
                           num_hidden = 50,
                           input_size = NULL,
                           num_embed = NULL, 
                           num_decode = 1,
                           masking = F, 
                           loss_output = "linear",
                           dropout = 0.2, 
                           ignore_label = -1,
                           cell_type = "lstm",
                           output_last_state = F,
                           config = "seq-to-one")

#graph.viz(symbol, type = "graph", direction = "LR", graph.height.px = 600, graph.width.px = 800)

# train.data <- mx.io.arrayiter(
#           data = matrix(rnorm(100, 0, 1), ncol = 20)
#           , label = rnorm(20, 0, 1)
#           , batch.size = 20
#           , shuffle = F
#                  )

train.x <- array(
               t(matrix(rnorm(100, 0, 1), nrow = 1))
               , dim = c(5, 20)
)

train.y <- matrix(rnorm(20, 0, 1), nrow = 1)

nn_model <- mx.model.FeedForward.create(
     symbol,
     X = train.x,
     y = train.y,
     ctx = mx.cpu(),
     begin.round = 1,
     num.round = 1000,
     optimizer = "sgd",
     learning.rate = 0.01,
     initializer = mx.init.uniform(0.01),
     eval.metric = mx.metric.mse,
     array.batch.size = 1,
     array.layout = 'colmajor'
)
```

Alexey

**Comment by Roshrini** [comment]
@alexmosc Thanks for trying it out.
For how-to and usability questions, please use MXNet discuss forum - https://discuss.mxnet.io/ to involve wider community.
We want to use Github issues mainly for bugs and feature request. Feel free to reopen if closed in error.

@sandeep-krishnamurthy Can you add labels: Question, RNN, R and close this issue?

**Comment by ankkhedia** [comment]
Hi @alexmosc Thanks for using MXNetR. I can help you with the question if I understand the exact use case and has some sort of sample dataset.
I want to understand what exactly you are trying to learn. When unroll config is set to seq-to-one , then there is one output for each sequence. So, if you have standard n_samples*seq_len training examples your label should be a vector of size n_samples.[Example- sentiment analysis http://dmlc.ml/rstats/2017/10/11/rnn-bucket-mxnet-R.html]

It works very differently when unroll config is set to one to one. Here training data arrays and training label arrays should exactly be of same size. In other words , the input as well as outputs are sequence and we learn on that data.[Example time series - https://jeremiedb.github.io/mxnet_R_bucketing/TimeSeries_CPU]

So, if you could give me a better idea of what exactly is your train data and labels are and what you exactly want to learn in the model I can try helping you.

@nswamy @sandeep-krishnamurthy Could your please tag the issue as Pending Requester Info?

**Comment by alexmosc** [comment]
Dear **ankkhedia**,

Thank you for replying to me.

My use case is the following (I will try to create a dummy dataset here in the body of message):

Let's say, batch size is 8
sequence length is 4
dimensions of timeseries equal 3.

Each sequence has one scalar label as output.

```
One sequence is
123
315
417
718
output: 1

Here starts sequence 2 from the same batch.
```

Now, with that having been said, a batch of data is:

```
X <- data.matrix(rnorm(96), ncol = 3)
y <- rnorm(8)
```

Starting from that point, I have a hard time coding the algorithm, and I would appreciate your help.

**Comment by ankkhedia** [comment]
Hi @alexmosc Thanks for getting back with the details. I am trying to look up existing tutorials/documentation for multidimensional data use case.

@jeremiedb Could you please point to any existing tutorial/documentation dealing with multidimensional time series for MXNetR or could provide some inputs on how the above use cases can be dealt?

**Comment by ankkhedia** [comment]
Hi @alexmosc 
Please find a small example training on random multidimensional data. I trained a small RNN for 5 epochs. This is just an example with random values and I didn't check for accuracy but just wanted a demonstration of how training could be done.

library("readr")
library("dplyr")
library("plotly")
library("mxnet")
seq_len = 100
samples = 192

seeds <- runif(samples, min = -pi, max = pi)
pts <- sapply(seeds, function(x) sin(x + pi/12 * (0:(seq_len))))

x <- pts[1:seq_len, ]
x <- array(x, dim = c(3, seq_len, samples/3))
y1 <- runif(32, 0.0, 1.0)
y2 <- runif(32, 0.0, 1.0)
batch.size = 8
## I have assumed 3 dimensional time series of seq_len 100 and having 64 samples 
train.data <- mx.io.arrayiter(data = x[,,1:32, drop = F], label = y1, 
                              batch.size = batch.size, shuffle = FALSE)

eval.data <- mx.io.arrayiter(data = x[,,-(1:32), drop = F], label = y2, 
                              batch.size = batch.size, shuffle = FALSE)

ctx <- mx.gpu()

initializer <- mx.init.Xavier(rnd_type = "gaussian", 
                              factor_type = "avg", 
                              magnitude = 3)

optimizer <- mx.opt.create("adadelta", rho = 0.9, eps = 1e-5, wd = 1e-6,
                           clip_gradient = 1, rescale.grad = 1/batch.size)


logger <- mx.metric.logger()
epoch.end.callback <- mx.callback.log.train.metric(period = 1, logger = logger)
batch.end.callback <- mx.callback.log.train.metric(period = 50)

symbol<- rnn.graph(config = "seq-to-one", cell_type = "lstm", 
                            num_rnn_layer = 1, num_embed = NULL, num_hidden = 4, 
                            num_decode = 1, input_size = NULL, dropout = 0.5, 
                            ignore_label = -1, loss_output = "softmax",
                            output_last_state = F, masking = T)

system.time(
  model <- mx.model.buckets(symbol = symbol,
                            train.data = train.data, 
                            eval.data = eval.data, 
                            num.round = 5, ctx = ctx, verbose = TRUE,
                            metric = mx.metric.accuracy, 
                            initializer = initializer, optimizer = optimizer, 
                            batch.end.callback = NULL, 
                            epoch.end.callback = epoch.end.callback)
)


Please let me know if this helps your use case.

**Comment by ankkhedia** [comment]
@sandeep-krishnamurthy @nswamy Could you please close this issue as the issue has been resolved

@alexmosc Feel free to reopen if closed in error
